rfcs/2022-10-12-14742-flexible-metric-tags.md

```diff
@@ -0,0 +1,306 @@
+# RFC 14742 - Support for Duplicate and Bare Tags on Metrics
+
+Vector's current metric model supports only a single value for each tag. Several sources, however,
+however, may send metrics that contain multiple values for a given tag name, or bare tags with no
+value. Vectors tag support should be enhanced to handle this data.
+
+## Context
+
+- [Support duplicate tag and bare tags on metrics](https://github.com/vectordotdev/vector/issues/14742)
+- [DataDog Source -> Filter Transform -> DataDog Sink dropping kube_service tag](https://github.com/vectordotdev/vector/issues/14707)
+- [Invalid custom_tags metrics from sinks datadog_metric](https://github.com/vectordotdev/vector/issues/14239)
+
+## Scope
+
+### In scope
+
+- Changes to the existing metric tag internal and external representations to support multiple
+  values and bare tags.
+
+### Out of scope
+
+- Changes to any other aspect of metric or event representations.
+
+## Pain
+
+When ingesting data from certain metric sources, the tags provided by that source may contain data
+that Vector's data model cannot currently represent. This is causing problems for prospective users
+that want to add Vector to their observability pipelines but cannot do so because it would cause
+data loss. Those sources include:
+
+- `datadog_agent` natively stores tags as bare strings, with optional values separated from the name
+  by a colon.
+- The Prometheus text encoding used by the `prometheus_scrape` source and `prometheus_exporter` sink
+  can support both repeated tag names and bare tags (without a value).
+- The Prometheus `remote_write` encoding similarly supports repeated tag names.
+- OpenTelemetry attributes support keys having arrays of values, which corresponds to repeated tags.
+
+## Proposal
+
+### User Experience
+
+#### External Representation
+
+The existing Protobuf representation of metric tags has a map of string-to-string pairs of
+tags. A new tags map will be added where the values are lists of optional strings, matching the
+semantics given above.
```

Comment:
OTLP also supports other primitive types as values in Attributes, is that something we would want to consider supporting with this change - or do we want to just stringify those?

Reply:
I think I'd prefer to just stringify those, to keep this simple, but that runs the risk of needing another tag model change down the road.

Reply:
> That's a good point, I neglected OTel allows for the values to be any value including arrays or nested objects. I wonder if that makes them different enough from "tags" where most sinks only allow strings, to put them in the metric event metadata and let users copy them into tags in a `remap` transform if they want (and can handle the different object types).

Checking the otel-contrib prometheusreceiver, Prometheus labels are stored as Attributes. By the same (reversed logic), when we're converting from `otlp` metrics to Vector's metric model we would put them into `.tags` today.